esp/services/ws_access/ws_accessService.cpp

```diff
@@ -2162,6 +2162,572 @@ bool Cws_accessEx::onResourcePermissions(IEspContext &context, IEspResourcePermi
+        resp.setResult("Successfully deleted a view.");
+    }
+    catch (IException* e)
+    {
+        FORWARDEXCEPTION(context, e, ECLWATCH_INTERNAL_ERROR);
+    }
+
+    return true;
+}
+
+bool Cws_accessEx::onQueryViewColumns(IEspContext &context, IEspQueryViewColumnsRequest &req, IEspQueryViewColumnsResponse &resp)
+{
+    try
+    {
+        CLdapSecManager* secmgr = queryLDAPSecurityManager(context);
+        
+        if(secmgr == NULL)
+            throw MakeStringException(ECLWATCH_INVALID_SEC_MANAGER, MSG_SEC_MANAGER_IS_NULL);
+
+        checkUser(context);
+        
+        IArrayOf<IEspViewColumn> viewColumns;
+        StringArray files, columns;
+
+        try
+        {
+            secmgr->queryViewColumns(req.getViewname(), files, columns);
+
+            ForEachItemIn(i, files)
+            {
+                Owned<IEspViewColumn> oneViewColumn = createViewColumn();
+                oneViewColumn->setViewname(req.getViewname());
+                oneViewColumn->setFilename(files.item(i));
+                oneViewColumn->setColumnname(columns.item(i));
+                viewColumns.append(*oneViewColumn.getLink());
+            }
+        }
+        catch (IException* e)
+        {
+            StringBuffer failedMsg("Failed: ");
+            StringBuffer eMsg;
+            resp.setSuccess(false);
+            resp.setResult(failedMsg.append(e->errorMessage(eMsg)));
+            return false;
+        }
+
+        resp.setSuccess(true);
+        resp.setResult("Successfully queried all view columns.");
+        resp.setViewcolumns(viewColumns);
+    }
+    catch (IException* e)
+    {
+        FORWARDEXCEPTION(context, e, ECLWATCH_INTERNAL_ERROR);
+    }
+
+    return true;
+}
+
+bool Cws_accessEx::onAddViewColumn(IEspContext &context, IEspAddViewColumnRequest &req, IEspAddViewColumnResponse &resp)
+{
+    try
+    {
+        CLdapSecManager* secmgr = queryLDAPSecurityManager(context);
+        
+        if(secmgr == NULL)
+            throw MakeStringException(ECLWATCH_INVALID_SEC_MANAGER, MSG_SEC_MANAGER_IS_NULL);
+
+        checkUser(context);
+
+        try
+        {
+            StringArray files, columns;
+
+            files.append(req.getFilename());
+            columns.append(req.getColumnname());
+
+            secmgr->addViewColumns(req.getViewname(), files, columns);
+        }
+        catch (IException* e)
+        {
+            StringBuffer failedMsg("Failed: ");
+            StringBuffer eMsg;
+            resp.setSuccess(false);
+            resp.setResult(failedMsg.append(e->errorMessage(eMsg)));
+            return false;
+        }
+
+        resp.setSuccess(true);
+        resp.setResult("Successfully added a view column.");
+    }
+    catch (IException* e)
+    {
+        FORWARDEXCEPTION(context, e, ECLWATCH_INTERNAL_ERROR);
+    }
+
+    return true;
+}
+
+bool Cws_accessEx::onDeleteViewColumn(IEspContext &context, IEspDeleteViewColumnRequest &req, IEspDeleteViewColumnResponse &resp)
+{
+    try
```

Comment:
I've made the exception mechanism more similar to other ws_access web services. Response no longer contain "success" or "result" fields, and now catch block is simply a one line:
`FORWARDEXCEPTION(context, e, ECLWATCH_INTERNAL_ERROR);`

Regarding queryLDAPSecurityManager, I looked at the option of creating a separate function that throws when queryLDAPSecurityManager fails.

However, given the purpose of queryLDAPSecurityManager (which returns NULL pointer when it is not available and let the caller decide what to do with NULL pointer), I felt that creating another function that wraps around queryLDAPSecurityManager, just to throw exception when it fails, was an overkill.